lib/Sema/TypeCheckSwitchStmt.cpp

```diff
@@ -372,10 +371,17 @@ namespace {
         PAIRCASE (SpaceKind::Constructor, SpaceKind::Constructor): {
           // Optimization: If the constructor heads don't match, subspace is
           // impossible.
+
           if (this->Head != other.Head) {
-            return false;
+            if (this->Head.getBaseIdentifier() !=
+                other.Head.getBaseIdentifier())
+              return false;
+            if (this->Head.getArgumentNames().size() != 0 &&
+                other.Head.getArgumentNames().size() != 0) {
```

Comment:
This sequence of checks doesn't make sense to me.  To preserve the optimization, we should be checking if the `baseIdentifier`s are different and the `argumentNames` differ in number.

Reply:
I may have misinterpreted some fo the tests.  I think this was attempting to take care of the case where if the case statement had no arguments/associated values (i.e. `.a`) compared to one where it (`.a(x:)`) then it should be considered a subspace.

Reply:
We pre-expand patterns via lookup to make that impossible.  Otherwise we wouldn't be able to detect overlapping patterns.

Reply:
Ah ok, makes sense